Installation and Usage documentation (#32)
univieCUBE/phenotrex: docs/usage.rst

Usage
=====

The following quick start example shows training, evaluation and inference with phenotrex models
using the phenotrex command line interface. For further information on flags and parameters used
in this tutorial, please consult the relevant CLI documentation available via
``phenotrex <command> --help``.

To illustrate phenotrex's capabilities, a small dataset of genome assemblies will be used.
To download all required data, run:

.. code-block:: console

    $ curl -OL http://fileshare.csb.univie.ac.at/phenotrex/tutorial_data.tar
    $ tar -xf tutorial_data.tar
    $ cd tutorial_data

To then install phenotrex, including its capability for extracting features from FASTA files,
in a new virtual environment:

> .. code-block:: console
> 
>     $ python3 -m venv env
>     $ source ./env/bin/activate
>     $ pip install phenotrex[fasta]

Reviewer: For users less familiar with venvs, it might be helpful to give more expressive names (what is the command, what is the name)
```suggestion
.. code-block:: console

    $ python3 -m venv phenotrex-env
    $ source ./phenotrex-env/bin/activate
    $ pip install phenotrex[fasta]
```

Creation of Phenotrex Input Features
------------------------------------
Phenotrex operates on presence/absence patterns of `eggNOG`_ cluster IDs in the passed genome.
If a DNA FASTA file is passed to phenotrex,
`Prodigal`_ is first used to find protein sequences - this step is skipped if a protein FASTA file

> is passed instead. To then find eggNOG cluster IDs from protein sequences, `deepnog`_ is utilized.

Reviewer: ```suggestion
is passed instead. To then find eggNOG cluster IDs from protein sequences, `deepnog`_ is used.
```
"utilize" has a slight connotation of "abuse"

Author: TIL!

Input files to feature creation may thus be DNA or protein multi-FASTA files, which may optionally
be gzipped.

Feature creation is computationally demanding. For this reason, direct input of individual
FASTA files to phenotrex is only implemented for prediction.
For model training, evaluation (and for batch prediction), tabular files must be created from
input FASTA files representing the genotype of all input files, encoded as eggNOG cluster IDs. This
allows reuse of the created features for all training and evaluation purposes.

To create a tabular genotype file suitable for use in phenotrex training:

.. code-block:: console

    $ phenotrex compute-genotype \
        --out T3SS.train_eval.genotype \

>         --n_threads 4 \

Reviewer: Iirc, this was recently renamed?
```suggestion
        --threads 4 \
```

        train_eval/genomes/*.fna.gz

After some time, this will create a new tab-separated values (TSV) file ``T3SS.train_eval.genotype``

> in the directory, of the following shape:

Reviewer: ```suggestion
in the current directory, of the following shape:
```

    supplied to the ``compute-genotype`` command (``--threads``). However, for large sets of genomes
    and when compute cluster resources are available to the user, it may be more expedient to compute
    subsets of genomes in parallel on different machines, and concatenate them afterwards.

Training of Phenotrex Classifiers
---------------------------------
To train a phenotrex classifier, two tabular input files are required: The ``genotype`` file
(created from FASTA files in the last section), containing representations of the input genomes; and
the ``phenotype`` file, containing true phenotypes for each input genome on which to train and
evaluate the model. For this tutorial, we provide a phenotype file containing information on Type 3
secretion system (T3SS) presence in each of the input genomes.

The tabular phenotype file required for training and model evaluation has the following shape:

.. code-block:: console

    Identifier	T3SS
    GCF_000012905.2.fna.gz	NO
    GCF_000195735.1.fna.gz	NO
    GCF_000060345.1.fna.gz	NO
    GCF_000220235.1.fna.gz	NO
    GCF_000190695.1.fna.gz	NO
    GCF_000015365.1.fna.gz	NO
    GCF_000173115.1.fna.gz	NO
    GCF_000173095.1.fna.gz	NO

>     GCA_003096415.1.fna.gz	NO

Reviewer: Should we include at least one `YES` sample?

Author: I was wondering already, I simply took the head of the file but you may be right

The first column of the file contains identifiers (file names) mapping to those in the genotype file,
and the second column contains true phenotypes. During training, the model will store the header of
column 2 as the name of the phenotype.

Phenotrex implements model training using two different machine learning algorithms:
`XGBoost`_ (XGB) and `Support Vector Machine`_ (SVM). For each algorithm, a number of hyperparameters
are settable for training and evaluation. Please consult the output of
``phenotrex train xgb --help`` and ``phenotrex train svm --help``, as well as the relevant documentation of the
underlying implementations. When no hyperparameters are selected, reasonable
(but possibly suboptimal) defaults are used.

To train an XGB classifier with the previously created genotype and the given phenotype file:

.. code-block:: console

    $ phenotrex train xgb \
        --genotype T3SS.train_eval.genotype \
        --phenotype train_eval/T3SS.train_eval.phenotype \
        --weights \
        --out T3SS.pkl

> This will create a new model artifact ``T3SS.pkl`` in the directory, and a

Reviewer: ```suggestion
This will create a new model artifact ``T3SS.pkl`` in the current directory, and a
```

tabular file ``T3SS.pkl.rank`` representing the relative impact of input features on prediction
output as learned by the model.

Performance Estimation of Phenotrex Classifiers
-----------------------------------------------
The default way for phenotrex to estimate model performance (other than applying the trained model
to a held back test set) is `nested cross-validation`_ (CV).
This allows the estimation of predictive performance for a given set of training data and hyperparameters.

To estimate performance of the model trained in the previous section, we perform a 10x/5x nested
cross-validation like so:

.. code-block:: console

    $ phenotrex cv xgb \
        --genotype T3SS.train_eval.genotype \
        --phenotype train_eval/T3SS.train_eval.phenotype \
        --out T3SS.misclassifications.tsv \
        --folds 5 \
        --replicates 10 \
        --threads 4

After training, predictive performance metrics averaged over outer CV folds will be printed
to stderr, and a new tabular file ``T3SS.misclassifications.tsv`` will be created. This file

> contains the identifiers, phenotype label and fraction of misclassifications of the sample over

Reviewer: ```suggestion
contains the identifiers, phenotype labels and fractions of misclassifications of the sample over
```

outer CV folds.

.. note::

    The above command does not accept a trained model artifact.
    Since cross-validation is performed by training several models on subsets of the given data, a final
    model is not warranted here.
    In general, training of the final classifier with ``phenotrex train {xgb,svm}``
    should be performed only when satisified with performance of the selected hyperparameters as given
    by cross-validation.

Performance Estimation for Metagenomic Phenotrex Classifiers
------------------------------------------------------------

> For phenotrex models intended to be applied to metagenomic bins, it is useful to estimate the impact

Reviewer: ```suggestion
For phenotrex models intended to be applied to metagenomic bins or metagenome assembled genomes, it is useful to estimate the impact
```
Seems like the more formal nomenclature to me, but feel free to ignore.

importance of that feature in the decision process of the model (irrespective of the final prediction),
when using SVM, the measure correlates with the probability of calling YES (positive
values) or NO (negative values) for the phenotype in question.

Feature Importance at Prediction Time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
A second, and arguably more useful type of explanation can be computed at prediction time.
For each predicted genome, a list of features is created which, either by presence or absence,
contributed most to the prediction output for that genome.
Feature importance is represented by `SHAP`_ (SHapley Additive exPlanations) values.
The sum of SHAP values of all features considered by the model is directly related to the probability
of calling YES for the phenotype and genome in question.

.. note::

    Feature explanation at prediction time is implemented by the `shap package`_, which efficiently
    computes the required explanations for XGB models with de facto zero overhead.
    For SVM models however, this calculation can be extremely costly. We thus suggest that for

>     use cases where model introspectability is important, XGB should be preferred over SVM.

Reviewer: Is that a word? My personal preference would be explainability or interpretability. Also, we are talking about explanations all the time.

Author: fair
